src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs

```diff
@@ -69,7 +70,8 @@ public int GetHashCode(BoundAttributeDescriptor descriptor)
                 }
             }
 
-            foreach (var metadata in descriptor.Metadata)
+            Debug.Assert(descriptor.MetadataImpl is not null, "We expect MetadataImpl to be configured in the default code path.");
```

Comment:
@NTaylorMullen thoughts on this? It looks like the JsonConverter uses the `DefaultTagHelperDescriptorBuilder` which will end up configuring this correctly. While we could fallback to using `Metadata`, it does make it harder to discover if this code is boxing the enumerator.

Reply:
Ya that's a fair concern. I feel like this may break other things like tests if they're not configured properly. Are we not able to attempt a cast on Metadata?